File: 38주차/여행경로.py
from collections import defaultdict
def solution(tickets):
    
    route = defaultdict(list)# 항공권 루트 정보를 담을 dictionary {'ICN':[],,}
    visited = defaultdict(list)# 모든 항공권을 다 써야하므로, 이를 확인할 visited 배열.
    n = len(tickets)
    for a,b in tickets:# dictionary 초기화.
        route[a].append(b)
        visited[a].append(0)
    for key in route:# 경로가 여러 개라면 사전순으로 출력해야하므로 미리 sort.
        route[key].sort()
    total = []
    
    def dfs(res,airport):
        if len(res) == n+1:
            total.append(res)
            return
        for i,next in enumerate(route[airport]):# 현재 공항에서 갈 수 있는 항공권 중에서
            if not visited[airport][i]:# 방문을 하지 않은 경로 선택
                visited[airport][i] = 1
                dfs(res+[next],next)# res 인자를 통해 경로를 저장하고, next:  그 다음 공항 인수로 전달.
                visited[airport][i] = 0 # 백트래킹
    dfs(['ICN'],'ICN')# ICN부터 출발.
    
    return total[0]
# 여행 경로 복습 ok

File: 47주차/여행경로.js
function solution(tickets) {
  var answer = [];
  let n = tickets.length;
  let route = {};
  let visited = {};

  for (let c of tickets) {
    let [a, b] = c;

    if (a in route) {
      route[a].push(b); // 현재 공항에서 갈 수 있는 도시 할당.
      visited[a].push(0); // 각각 티켓의 방문 여부를 나타내는 배열.
    } else {
      route[a] = [b];
      visited[a] = [0];
    }
  }

  for (let c in route) route[c].sort(); // 경로가 여러개일때, 사전순 1개 고르기 위한 정렬.
  console.log(route);
  function dfs(now, res) {
    if (res.length === n + 1) {
      answer.push(res);
      return;
    }
    // 더이상 다음 경로로 갈 수 없는 경우, 딕셔너리에 키가 없어서 발생하는 에러를 막기 위함.
    if (now in route) {
      for (let i = 0; i < route[now].length; i++) {
        let city = route[now][i];
        if (!visited[now][i]) {
          visited[now][i] = 1;
          dfs(city, [...res, city]);
          visited[now][i] = 0;
        }
      }
    }
  }
  dfs("ICN", ["ICN"]);
  // console.log(answer)
  return answer[0];
} // 프로그래머스 dfs.
